### PR TITLE
feat: support csharp eager mode batch run

### PR DESCRIPTION
**Pull request by elliotzh**
# Description

This pull request primarily involves changes that enhance language support and improve the handling of flow schemas in the `promptflow` package. The changes include the introduction of validation and default values for the `language` field in flow schemas, modifications to the handling of eager flows in the batch engine, and the addition of methods to fetch flow metadata and input definitions.

Language Support and Flow Schema Changes:

* [`src/promptflow/promptflow/_sdk/schemas/_flow.py`](diffhunk://#diff-84af29d72d79c61d55daa91de77a03235646967bcaa7c3299f33adcead97bc53R4-R8): The `language` field in the `BaseFlowSchema` class now has a default value and validation to ensure it's either Python or CSharp. In the `EagerFlowSchema` class, the `path` field is no longer required, and a `post_load` method was added to infer the path from the entry depending on the language. [[1]](diffhunk://#diff-84af29d72d79c61d55daa91de77a03235646967bcaa7c3299f33adcead97bc53R4-R8) [[2]](diffhunk://#diff-84af29d72d79c61d55daa91de77a03235646967bcaa7c3299f33adcead97bc53L42-R47) [[3]](diffhunk://#diff-84af29d72d79c61d55daa91de77a03235646967bcaa7c3299f33adcead97bc53L61-R84)

* [`src/promptflow/promptflow/azure/_entities/_flow.py`](diffhunk://#diff-b8b37b2a9593a8fb1f0b078663aad36d8b4009a1205bd4e799a6180d666b7e65R16): Added a `language` property method to fetch the language of the flow. [[1]](diffhunk://#diff-b8b37b2a9593a8fb1f0b078663aad36d8b4009a1205bd4e799a6180d666b7e65R16) [[2]](diffhunk://#diff-b8b37b2a9593a8fb1f0b078663aad36d8b4009a1205bd4e799a6180d666b7e65R216-R219)

Handling of Eager Flows in Batch Engine:

* [`src/promptflow/promptflow/batch/_batch_engine.py`](diffhunk://#diff-ecf0905f2116abe08b5cf2931b856bf39aa8b38a30fadd9042538d076dbfde80L13-R13): Several changes were made to handle eager flows differently. The `__init__` method now checks if the flow is eager and its language. The `run` and `_exec` methods have been adjusted to handle eager flows. The `_update_aggr_result` method has been modified to skip eager flows. [[1]](diffhunk://#diff-ecf0905f2116abe08b5cf2931b856bf39aa8b38a30fadd9042538d076dbfde80L13-R13) [[2]](diffhunk://#diff-ecf0905f2116abe08b5cf2931b856bf39aa8b38a30fadd9042538d076dbfde80L102-L109) [[3]](diffhunk://#diff-ecf0905f2116abe08b5cf2931b856bf39aa8b38a30fadd9042538d076dbfde80L176-R175) [[4]](diffhunk://#diff-ecf0905f2116abe08b5cf2931b856bf39aa8b38a30fadd9042538d076dbfde80L248-R245) [[5]](diffhunk://#diff-ecf0905f2116abe08b5cf2931b856bf39aa8b38a30fadd9042538d076dbfde80L339-R336) [[6]](diffhunk://#diff-ecf0905f2116abe08b5cf2931b856bf39aa8b38a30fadd9042538d076dbfde80L404-R410)

Fetching Flow Metadata and Input Definitions:

* [`src/promptflow/promptflow/batch/_base_executor_proxy.py`](diffhunk://#diff-44ac8eeb30a630bd24987e92f8bd5a180d57a3ad067db5d937561f21771e14c5R33-R49): Added `_get_flow_meta` and `get_inputs_definition` methods to fetch flow metadata and input definitions, respectively.

* [`src/promptflow/promptflow/batch/_csharp_executor_proxy.py`](diffhunk://#diff-f8219170007d8a89eed104a166ab8c0c9d3af5b5c5e61225659d982f470aeb95R5-R15): Added `_get_flow_meta` and `_generate_flow_meta` methods to fetch and generate flow metadata, respectively. The `create` method has been modified to generate flow metadata for eager flows. [[1]](diffhunk://#diff-f8219170007d8a89eed104a166ab8c0c9d3af5b5c5e61225659d982f470aeb95R5-R15) [[2]](diffhunk://#diff-f8219170007d8a89eed104a166ab8c0c9d3af5b5c5e61225659d982f470aeb95L22-R77) [[3]](diffhunk://#diff-f8219170007d8a89eed104a166ab8c0c9d3af5b5c5e61225659d982f470aeb95R93-R118) [[4]](diffhunk://#diff-f8219170007d8a89eed104a166ab8c0c9d3af5b5c5e61225659d982f470aeb95L63-R134)

Other Changes:

* [`src/promptflow/promptflow/_sdk/operations/_local_storage_operations.py`](diffhunk://#diff-1a5c2463c2ef997c5203a6b2e77b247d812961b7a6df5f0f45c02579b6656719L17): The import of `load_flow` has been moved to a different location. [[1]](diffhunk://#diff-1a5c2463c2ef997c5203a6b2e77b247d812961b7a6df5f0f45c02579b6656719L17) [[2]](diffhunk://#diff-1a5c2463c2ef997c5203a6b2e77b247d812961b7a6df5f0f45c02579b6656719R26)

* [`src/promptflow/promptflow/batch/_csharp_executor_proxy.py`](diffhunk://#diff-f8219170007d8a89eed104a166ab8c0c9d3af5b5c5e61225659d982f470aeb95R149-R150): The `destroy` method has been updated to delete temporary DAG files if they exist.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/batch/_batch_engine.py [hunk 1]
```diff
@@ -160,8 +159,8 @@ def run(
                 try:
                     # register signal handler for python flow in the main thread
                     # TODO: For all executor proxies that are executed locally, it might be necessary to
-                    # register a signal for Ctrl+C in order to customize some actions beyond just killing
-                    # the process, such as terminating the executor service.
+                    #  register a signal for Ctrl+C in order to customize some actions beyond just killing
+                    #  the process, such as terminating the executor service.
```
Comment (PeiwenGaoMS) [inline src/promptflow/promptflow/batch/_batch_engine.py:163]: Extra space

Comment (elliotzh) [reply src/promptflow/promptflow/batch/_batch_engine.py:163]: this is still the TODO sentence, so need an extra space to let IDE know this
image
image

File: src/promptflow/promptflow/batch/_batch_engine.py [hunk 2]
```diff
@@ -99,14 +99,13 @@ def __init__(
         """
         self._flow_file = flow_file
         self._working_dir = Flow._resolve_working_dir(flow_file, working_dir)
-        if self._is_eager_flow_yaml():
-            self._is_dag_yaml_flow = False
-            self._program_language = FlowLanguage.Python
-        else:
+
+        self._is_eager_flow, self._program_language = self._check_eager_flow_and_language_from_yaml()
```
Comment (elliotzh) [inline src/promptflow/promptflow/batch/_batch_engine.py:103]: @lumoslnt we don't have an official name for non-eager flow, so I think `_is_eager_flow` is a more clear name

Comment (lumoslnt) [reply src/promptflow/promptflow/batch/_batch_engine.py:103]: Good to me.

File: src/promptflow/promptflow/batch/_batch_engine.py [hunk 3]
```diff
@@ -401,11 +398,13 @@ def _update_aggr_result(self, aggr_result: AggregationResult, aggr_exec_result:
...
-            flow_file = self._working_dir / self._flow_file if self._working_dir else self._flow_file
-            with open(flow_file, "r", encoding="utf-8") as fin:
-                flow_dag = load_yaml(fin)
-            if "entry" in flow_dag:
-                return True
-        return False
+    def _check_eager_flow_and_language_from_yaml(self):
+        flow_file = self._working_dir / self._flow_file if self._working_dir else self._flow_file
+        # TODO: remove this after path is removed
+        if flow_file.suffix == ".dll":
```
[the remark this answers is not in the document]
Comment (elliotzh) [reply src/promptflow/promptflow/batch/_batch_engine.py]: good catch, updated

File: src/promptflow/promptflow/_sdk/schemas/_flow.py [hunk 4]
```diff
@@ -58,6 +63,22 @@ class EagerFlowSchema(BaseFlowSchema):
     """Schema for eager flow."""
 
     # path to flow entry file.
-    path = LocalPathField(required=True)
+    path = LocalPathField(required=False)
     # entry function
     entry = fields.Str(required=True)
+
+    @post_load
+    def infer_path(self, data: dict, **kwargs):
```
Comment (D-W-) [inline src/promptflow/promptflow/_sdk/schemas/_flow.py:71]: Why not use `@validates_schema` as offline synced?
https://marshmallow.readthedocs.io/en/latest/extending.html#schema-level-validation

**Review by lumoslnt — DISMISSED** [review]

File: src/promptflow/promptflow/_sdk/schemas/_flow.py (hunk 4 above)
Comment (elliotzh) [reply src/promptflow/promptflow/_sdk/schemas/_flow.py:71]: here the major purpose is to infer path before your PR merged. Will update it after your PR merged.

File: src/promptflow/promptflow/_sdk/schemas/_flow.py [hunk 5]
```diff
@@ -58,6 +63,22 @@ class EagerFlowSchema(BaseFlowSchema):
...
+        # TODO: remove this after path is removed
+        language = data.get(LANGUAGE_KEY, FlowLanguage.Python)
+        if language == FlowLanguage.Python and data.get("path", None) is None:
+            raise ValidationError(message={"path": ["Missing data for required field."]})
+        elif language == FlowLanguage.CSharp and data.get("path", None) is None:
+            # for csharp, path to flow entry file will be a dll path inferred from
+            # entry by default given customer won't see the dll on authoring
+            m = re.match(r"\((.+)\)[a-zA-Z0-9]+(\.[a-zA-Z0-9]+)+", data["entry"])
+            if not m:
+                raise ValueError(f"Invalid entry: {data['entry']}")
```
Comment (D-W-) [inline src/promptflow/promptflow/_sdk/schemas/_flow.py:82]: Why not validation error. @Stephen1993 what's the error category for value error?

**Review by D-W- — APPROVED** [review]

**Review by lumoslnt — APPROVED** [review]

**Review by wangchao1230 — APPROVED** [review]

**Review by D-W- — APPROVED** [review]

**Review by wangchao1230 — APPROVED** [review]

**Review by D-W- — APPROVED** [review]

**Review by PeiwenGaoMS — APPROVED** [review]

**Review by lumoslnt — APPROVED** [review]

**Pull request merged**